src/app/calendar-app/runbox-calendar-event.ts

```diff
@@ -69,137 +88,437 @@ export class RunboxCalendarEvent implements CalendarEvent {
-        if (this.ical.getFirstSubcomponent('vtimezone')) {
-            for (const tzComponent of this.ical.getAllSubcomponents('vtimezone')) {
-                const tz = new ICAL.Timezone({
-                    tzid:      tzComponent.getFirstPropertyValue('tzid'),
-                    component: tzComponent,
-                });
+        // Not sure how we get "no icalevent.component", came up in tests tho
+        if (icalevent.component && icalevent.component.parent) {
+            this.ical = icalevent.component.parent;
+        } else {
+            this.ical = new ICAL.Component('vcalendar');
+            this.ical.addSubcomponent(icalevent.component);
+        }
+        this.event = icalevent;
+        // starttime of this particular instance! if recurring, not
+        // necessarily the same as event.dtstart
+        this._dtstart = startdate;
+        this._dtend = enddate;
+    }
 
-                if (!ICAL.TimezoneService.has(tz.tzid)) {
-                    ICAL.TimezoneService.register(tz.tzid, tz);
+    // An exception to the main (recurring) event
+    addExceptionEvent(
+        origdate: moment.Moment,
+        startdate: moment.Moment,
+        enddate: moment.Moment,
+        thisandfuture: boolean,
+        title: string,
+        description: string,
+        location: string): boolean {
+        if (this.isException) {
+            // This shouldnt be possible
+            console.log('Refusing to create an exception of an exception');
+            return false;
+        }
+        // clone existing one
+        const new_exception = new ICAL.Event(ICAL.Component.fromString(this.event.toString()));
+        new_exception.component.removeProperty('rrule');
+        const recurrence_id = this.momentToIcalTime(origdate, this.event.startDate.zone);
+        const new_start = this.momentToIcalTime(startdate, this.event.startDate.zone);
+        let new_end;
+        if (enddate) {
+            new_end = this.momentToIcalTime(enddate, this.event.startDate.zone);
+        }
+        if (this._dtstart.isDate) {
+            recurrence_id.isDate = true;
+            new_start.isDate = true;
+            new_end.isDate = true;
+        }
+        new_exception.recurrenceId = recurrence_id;
+        if (thisandfuture) {
+            const rId = new_exception.component.getFirstProperty('recurrence-id');
+            rId.setParameter('range', 'THISANDFUTURE');
+        }
+        new_exception.startDate = new_start;
+        if (new_end) {
+            new_exception.endDate = new_end;
+        }
+        if (title) {
+            new_exception.summary = title;
+        }
+        if (description) {
+            new_exception.description = description;
+        }
+        if (location) {
+            new_exception.location = location;
+        }
+        this.ical.addSubcomponent(new_exception.component);
+        // regenerate this.event to sort out the exceptions/recurrences?
```

Comment:
Looks like this always returns false: perhaps it doesn't need a return type at all?

Reply:
you mean addExceptionEvent ? it doesn't have one *confused*

Reply:
It does on my screen :)

```
    addExceptionEvent(
        origdate: moment.Moment,
        startdate: moment.Moment,
        enddate: moment.Moment,
        thisandfuture: boolean,
        title: string,
        description: string,
        location: string): boolean {
        if (this.isException) {
            // This shouldnt be possible
            console.log('Refusing to create an exception of an exception');
            return false;
        }
```

Reply:
Ah! Oops